### DISCUSSION TITLE
Dynamically update Combo Box value

### DISCUSSION THREAD

**Discussion by mdurchholz** [uswds/uswds#5966] [Q&A]
Is there a simple way to dynamically update a combo box value? I am trying to update a combobox value based off the selection of  another combo box. I have a desktop and mobile version of a combobox and I would like them to sync together whenever an option is selected.

**Comment by mejiaj** [comment]
Hello, @mdurchholz!

Would you be able to provide a quick demo or codepen so that we can better understand this request?

**Reply by mdurchholz** [reply]
@mejiaj - After about a year, I finally have the time to get back into this. See my Codepen below:

https://codepen.io/mdurchholz/pen/KwdEBRz

Scenario 1 - Change combobox value
Scenario 2 - Change combobox menu items

**Reply by mejiaj** [reply]
@mdurchholz no worries and thanks for the pen. You'd have to put in a [feature request](https://github.com/uswds/uswds/issues) to the USWDS team. It'd help if you state your use case.

A hacky workaround might be to change the the list values and re-init the combobox, but that might not be the greatest user experience 🤔.

**Reply by mdurchholz** [reply]
@mejiaj - Do you have a recommendation on how to "reinit" the combobox? I don't see any function to do so in the documentation. I could be missing it, but I have only been able to init the combobox on page load when the script targets the USA classes.

**Reply by mejiaj** [reply]
@mdurchholz every component should have an `.on()` and `.off()` method. 

There's also other methods available you can see in [usa-combo-box/src/index.js](https://github.com/uswds/uswds/blob/9999d08fe6c8e7f0aaa3f24b86ba37b281c25b54/packages/usa-combo-box/src/index.js#L892), but by default it should have on/off.

Alternatively, have you tried dynamically creating two elements, one inside `fruit-desktop--list` (the enhanced list) and another in `.usa-select.usa-combo-box__select` (the select input where the list is created from)?

**Reply by mdurchholz** [reply]
@mejiaj - Can you point out which lines you are referring to in the Javascript file that contain the .on and .off methods? I am not seeing anything related to those functions in the index file you linked to.

Scenario 1:
The only way I have been able to address this scenario is by triggering clicks on the option elements. I grab the index of the selected combobox option and then trigger a click on the synced combobox list items that match that index value.

Scenario 2:
I have not begun to fully work on the menu change, but I have noticed that you can add/remove option elements to the select element, toggle the menu, and the menu options will rebuild based off the the select options. The issue there is that I have not found a way to programmatically open/close the menu from an outside,custom script. I have considered your idea of manually rebuilding the necessary select options and menu list elements tho.

**Reply by mejiaj** [reply]
@mdurchholz it needs documentation, but every component should use [behaviors](https://github.com/uswds/uswds/blob/9999d08fe6c8e7f0aaa3f24b86ba37b281c25b54/packages/uswds-core/src/js/utils/behavior.js#L116) and the behavior library has `on/off` methods for components. IIRC, the `.init()` method basically calls `.on()`.

**Reply by mdurchholz** [reply]
@mejiaj - How would you call these methods from an outside vanilla JS script? I see the target element is the parameter. Is it something like uswds.on(targetElement)?

**Reply by mejiaj** [reply]
@mdurchholz it's briefly documeted on the [docs for developers](https://designsystem.digital.gov/documentation/developers/#js-customization-2) page.

I would try adding the element dynamically to both lists I mentioned in this [comment](https://github.com/uswds/uswds/discussions/5966#discussioncomment-14342339) before using on/off though.

**Comment by mdurchholz** [comment]
@mejiaj - Does this documentation assume that we are using a library/framework like React and Angular? I don't see any documentation for regular Vanilla Javascript.

**Reply by mejiaj** [reply]
You'll need something like vite or esbuild to be able to import the JS file, but yes it works. You can see an example here:

```js
// src/assets/project/js/index.js

import usaModal from "@uswds/uswds/js/usa-modal";

(function () {
  document.addEventListener("DOMContentLoaded", () => {
    const myProjectModal = document.querySelectorAll(".usa-modal");

    usaModal.on(myProjectModal);
  });
})();
```

https://github.com/uswds/uswds-sandbox/commit/8be3844fcf1a891d8fccac52c39e1353d938b397#diff-d2da3b83bdac72353269b1268761339716dfe40887f42132bb39808bf49fb681

**Comment by ryparker** [comment]
Vanilla JavaScript works with the component modules; React or Angular is not required. With a bundler, the combo-box import is:

```js
import comboBox from "@uswds/uswds/js/usa-combo-box";

const root = document.querySelector("#filters");
comboBox.on(root);
```

Pass a DOM element containing the components, not the `NodeList` returned by `querySelectorAll`. Avoid also auto-initializing those same components with the full bundle.

I checked both update scenarios against USWDS 3.14.0:

- **Changing the selected value:** assigning the underlying select's `.value`, even with a `change` event, does not synchronize the visible input. There is no exported `setValue()` method.
- **Adding options:** the generated list reads the underlying select's options when it opens. Update the select's options, rather than manually maintaining both the select and generated list. If the list is already open, the exported `comboBox.displayList(comboElement)` rebuilds it. Removing or replacing the current selection still requires handling the selected value and displayed text together.

The earlier reinitialization suggestion needs correcting: `.off()` removes listeners, and `.on()` skips rebuilding an already-enhanced combo box. That sequence does not reset its value or recreate its markup.

For the desktop/mobile use case, one combo box with a responsive layout avoids keeping two independent selection states synchronized. Two synchronized instances would currently need an application-level adapter, not just `.on()`/`.off()`.
